### ISSUE TITLE
results load slow!

### ISSUE THREAD

**Issue by nevion** [KDAB/hotspot#97]
Using 1.1 AppImage on Ubuntu LTS 16.04.  The datacollect was fast, and yielded 500MB for me for about 10 seconds.  However the loading of results is terribly slow.  I've still yet to load a file.  Htop shows hotspot-perfparser is the CPU hog. 

Is the appimage a debug build?  Is this slow load time expected?

At 500 megabytes, shouldn't a binary format like this be parsed in about 5 seconds?  Finally loaded after about 20 minutes (nice to see it, good ui)

**Comment by milianw** [comment]
Hey there, sorry for the long delay in responding.

To answer your qeustion: No, the app image is not a debug build. No, the slow load time is _not_ expected.

Can you make this report more useful by giving me something to work with? Instructions to reproduce the issue? Or profile hotspot (yes, a bit meta, but doable) and show me the results you obtain there.

Thanks

**Comment by nevion** [comment]
Here you go.  Without reviewing the sources it looks like the dwarf library is too slow to test these things and that you could benefit by extracting all the metadata into local (optimized) datastructures to kill the expense, essentially caching to avoid hitting the dwarf apis.  I profiled the appimage's hotspot-perfparser on --input perf.data.myapp --max-frames 1024

![image](https://user-images.githubusercontent.com/2229688/31571557-fbe6e9dc-b049-11e7-9673-fd9a7bd8ee65.png)
![image](https://user-images.githubusercontent.com/2229688/31571561-03667ae2-b04a-11e7-98a7-73d941afad1b.png)

**Comment by nevion** [comment]
Looking at the source I think it could be worth trying a memoizing technique on perfsymboltable.cpp::find_die

That is, store the die in a map/LRU cache with the key being some combination of mod, die, and bias.  This is assuming these calls are being called these same mod, addr, and bias' of course.

**Comment by milianw** [comment]
First of all: Are you using Clang? Or are your binaries stripped of debug information? Either would explain why we are running into this code path in the first place.

Secondly, there is a cache already. So the question now becomes whether the cache is simply not working for your use-case (i.e. low cache hit ratio due to immense numbers of IPs encountered). Or, alternatively and actually more likely, whether something else goes wrong and makes us believe we have to clear the cache.

One way or another, I will need a proper way to reproduce this issue. If this is reproducible for you, can you tell me how? I.e. can you try to use `perf` and `hotspot` on a FOSS project that I can compile and run myself? That would allow me to inspect this issue in more detail and actually fix the issue(s) at hand. Otherwise, someone (you?) will have to find the answers to the questions above, and write patches for them.

**Comment by milianw** [comment]
Actually, I just read about `clang -gdwarf-aranges`. If you are using clang, can you recompile your software with this switch and see if it fixes the performance issue for you?

**Comment by nevion** [comment]
Was using gcc 5.4, stock ubuntu 16.04

There's only 500MB of data the parser parses 20 minutes is unrealistically slow and vtune can do this same sort of profiling/load in under a minute.

Part of my program was compiled with debug where as the libraries I linked against are not (static libs).

**Comment by nevion** [comment]
Can you indicate where the caching happens that you mentioned?

An easy (Not necessarily correct) explination for what happens could be check the expensive path because we don't have a cached result (because the value doesn't exist), and we redo this check many times.  Perhaps we need to store that we've tried this key as well.

**Comment by milianw** [comment]
To make it clear: Yes, using 20min for parsing 500MB is far too slow - there clearly is something going wrong on your system. On my system, I don't see this issue, I easily parse a few GB of `perf.data` in under one minute on my laptop.

This is why I need you to document the steps you take, so that I can reproduce this issue myself. The steps should include what program you are using and how you compile it, the workload you used for the program and the perf command you used.

Regarding where the cache is: have a look at `m_addressCache` in the `PerfSymbolTable`. Note that it can get dirty and thus needs to be cleared. It may indeed be that we fail to cache a negative result here. Or the cache gets cleared too often. I also wanted to split the cache up to make store the cache within the `ElfInfo` struct, which should allow us to keep the data around for longer.

**Comment by nevion** [comment]
Unfortunately the circumstance is that we'll have to be sparse on point 2.  The perf command is dispatched unaltered via hospot's record.  Supporting libraries are compiled without debug info, the main program is with profiling info (-g).

I'll run experimental builds if you want to give something a try; otherwise if I have time I will check to see if it is indeed the negative cache result.

**Comment by farmboy0** [comment]
I see also slow load times in hotspot.
I use GCC 5.4.0 as well (Gentoo stable).

**Comment by milianw** [comment]
@nevion sorry, I don't understand the first two sentences of your last comment. Can you rephrase them please? What is point 2?

@farmboy0 can you supply any of the information I requested?

**Comment by farmboy0** [comment]
Here is my ebuild for installing hotspot:
https://github.com/farmboy0/portage-overlay/tree/master/dev-util/hotspot
I think I tried to analyze a wine process.
I'll get you more info.

**Comment by milianw** [comment]
Also, could you guys run this on your `perf.data` files so we get an idea of how many unique IPs were encountered:

    perf script -i perf.data -F ip | sort | uniq | wc

**Comment by nevion** [comment]
~9650 for the file I reported on.

**Comment by milianw** [comment]
OK, that's not a lot at all. But I think I have finally found a way to reproduce this issue! I'll have a look at whether it can be fixed easily.

**Comment by milianw** [comment]
Guys, can you try https://transfer.sh/cbXlA/Hotspot-git.9039e50-x86_64.AppImage and see whether it helps with your situation?

In the scenario I encountered, this does bring down the parse speed from ~55s down to ~16s in one case. While this helps, it's still not ideal. It seems we still suffer from some bad handling of the `mmap` events, which leads us to clear the cache quite often...

**Comment by nevion** [comment]
@milianw good improvements, now it's down to 6m12s for the file I profiled previously.  Still several times too slow but that's a big difference in practical time already.  I'll do a reprofile (flamegraph)  via hotspot soon to help guide further advancements.

The command I used for this part was:
`time /tmp/.mount_HotspoYizNj8/usr/lib/x86_64-linux-gnu/libexec/hotspot-perfparser --input perf.data.myapp --max-frames 1024 2>&1 > /dev/null`

Another thing I was thinking is perhaps too much data is being written out to stdout, or isn't buffered sufficiently (e.g. flushed too often).  So I redid a timing:

`time /tmp/.mount_HotspoYizNj8/usr/lib/x86_64-linux-gnu/libexec/hotspot-perfparser --input perf.data.myapp --max-frames 1024|wc -l`

Which resulted in the following:
```
26182

real    18m3.881s
user    18m2.752s
sys     0m1.008s
```

Not sure if the overhead of an additional pipe/wc can explain getting back to the original timing but as I said, I'll rerun it with hotspot's profiling soon.

**Comment by nevion** [comment]
![hotspot-parser-flame2](https://user-images.githubusercontent.com/2229688/32205456-90488af8-bdac-11e7-9d55-cc1a60a24d51.png)

Ok, here's  the re-profile flamegraph of the parser parsing the original  datafile.

**Comment by milianw** [comment]
ok, so what's left is apparently the clang fallback. @nevion as I wrote above, I will need more information on how to reproduce this. I.e. some open source code project that I can compile the way you do, i.e. explicit steps for me to reproduce everything. From compiling, over running perf, until finally we run hotspot.

**Comment by nevion** [comment]
@milianw are there statistics, things like cache clears/misses you could add as more insightful diagnostics into the problem?  Would tracking particular call counts to the dwarf library functions also be something to add?  And do you suspect the dwarf library itself is slow?  The individual library functions do seem to add up to the majority of the function's call duration and if that's less than 1:1 then your caching is working and it's the underlying die parser that's the problem.

Have you heard of [libelfin](https://github.com/aclements/libelfin) before?  It's built on std unordered_map/vector and a small_vector (hybrid static_vector and regular vector).  Might be worth a spin to see if there's a drastic performance difference caused by high quality datastructures.

**Comment by milianw** [comment]
You can add those statistics, it's a simple two-liner probably.

I have not heard of libelfin before, will have a look.

But can you _please_ acknowledge the question I have asked multiple times now? Even if it's a "the code I reproduce this issue on is closed source" it will at least tell me that we'll need to find some other way to reproduce this...

**Comment by milianw** [comment]
Looking at `libelfin`, it doesn't seem to be usable currently. No API for unwinding e.g. I'm also very reluctant to switch to anything other than elfutils, considering this is complex stuff that has tons of corner cases. Accuracy and correctness is much more important than performance.

**Comment by nevion** [comment]
@millianw  I've mentioned previously about circumstance, and my lack of commenting on the matter further is in answer in of itself.  There's nothing to be gained focusing on that though.  I don't think I have a freak case of software here and you've got only a few early adopters from what I can tell - I hope we can move past these factors and improve the usability of hotspot as it's a polished and nice profiler front-end.  I'll see if I can give you a similar performance issue with profiling firefox, soon.

Part of my issue with hacking at hotspot/perfparser already has been the library requirements of hotspot (same reasons why AppImage build is of value), perhaps that wont' be so bad on perfparser standalone as Qt 5 isn't hard at all to get working on a system through the official tarball.  I don't think it'll be two-liner (I can get pendantic and say it'll be minimum 15 or 20, with particular placement being key such that 2-liner is misleading of effort/time) and on my end it's the time investment to do that and dev environment while I juggle other things.  I also want to be mindful of your time in support around the issue and thus far which has been something we could know more about factors by laying in place performance diagnostics that validate assumptions - less good than having the reproducability in-hand but a very good next best thing.  I see no problems reporting on statistics/diagnostics.

As for libelfein - my intent was 2 fold - one see if you knew about it and flag issues, but second to indicate it maybe worth an experiment - through whatever (hacky) means necessary and library pitfals, say not supporting unwinds, to characterize the perf difference of the two implementations on the same data.  Certainly accuracy/correctness first but assessing performance issues and what can be gained is also valuable.  Guess I should profile with libdw with debug symbols too though, on the same point.  My suspicions is that the C datastructures in use are poor...   their workhorse hashmap is this: https://sourceware.org/git/?p=elfutils.git;a=blob_plain;f=lib/dynamicsizehash.c;hb=ab6b37ac32ddf2f2f11f800a770170814f5cbb8b - which has a strange hash function (search for lookup) and cites Knuth.

**Comment by milianw** [comment]
So the answer is no, you cannot give me anything to reproduce this easily. Considering I have no known way to reproduce this, I have very little interest in spending time on this. If anyone else can add more useful information, that would be welcome.

I can add some stats in a temp branch eventually for you.

libelfein is out of question, not doing unwinding will remove the bottleneck of course, since you won't trigger the code path in question then. Also please understand the code before claiming a hashmap in it is to blame. That is very far fetched and completely besides the point.

**Comment by farmboy0** [comment]
$ perf script -i perf.data -F ip | sort | uniq | wc
Failed to open /lib64/libc-2.23.so, continuing without symbols
Failed to open /lib64/libpthread-2.23.so, continuing without symbols
Failed to open /lib64/ld-2.23.so, continuing without symbols
[vdso] with build id 60b133ad23b2e1bcefd4dd9cc99adf549c09d1ba not found, continuing without symbols
Failed to open /lib64/libm-2.23.so, continuing without symbols
   5701    5700  102601

**Comment by farmboy0** [comment]
unwind: target platform=x86 is not supported
Failed to open /tmp/perf-4970.map, continuing without symbols
Failed to open /lib32/libpthread-2.23.so, continuing without symbols
Failed to open /lib32/libc-2.23.so, continuing without symbols
Failed to open /mnt/daten/Wine/ff11/drive_c/PlayOnline/SquareEnix/PlayOnlineViewer/pol.exe, continuing without symbols
Failed to open /lib32/libm-2.23.so, continuing without symbols
Warning:
Processed 274769 events and lost 70 chunks!

Check IO/CPU overload!

  26810   26809  482563

**Comment by farmboy0** [comment]
Here's screenshot.
[Link](http://wikisend.com/download/148130/Screenshot_20171104_203804.jpeg)

**Comment by milianw** [comment]
guys, please test https://transfer.sh/LgKrW/Hotspot-git.5292768-x86_64.AppImage and see if it helps with your situation, too

**Comment by farmboy0** [comment]
This is alot faster with the provided appimage. It loads my perf.data in around 30s.

**Comment by nevion** [comment]
Loads fast - time of < 30 seconds for the perf.data.myapp I've used since the start.

**Comment by milianw** [comment]
OK, that's good to hear. Seems like the issues you guys have been seeing with GCC is somehow related to what I fixed for clang. I still wonder how exactly this goes wrong there, supposedly it comes from stripped binaries, or binaries without debug symbols - at least that's my hunch. Anyhow, since this seems to be resolved now, I'll close this report then. If there's anything else that needs to be optimized, please report a new task with concrete steps to reproduce.

**Comment by Marwes** [comment]
Any chance of a new release which includes the improvements made here?

**Comment by milianw** [comment]
Yes @Marwes, a new hotspot release is planned - I just need to find the time for that!

Please stay tuned for now and either build from source or download the latest AppImage from the travis CI daily cron job.
